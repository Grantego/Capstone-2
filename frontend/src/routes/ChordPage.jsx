import React from "react"
import { Outlet } from "react-router"
import SpellerApi from "../../api"
import NavBar from "../components/NavBar"
import AutoFillBar from "../components/AutoFillBar"
import { useLoaderData } from "react-router"

export async function loader() {
    SpellerApi.token = localStorage.getItem("token")
    const res = await SpellerApi.getAllChords()
    return res
}

const ChordPage = () => {
    const {chords} = useLoaderData()
    return (
        <>
            <NavBar/>
            <AutoFillBar chords={chords.map(chord => chord.name)}/>
            <div id="chord-details"><Outlet/></div>
        </>
    )
}


export default ChordPage