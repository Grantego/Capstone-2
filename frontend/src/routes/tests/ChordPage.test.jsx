import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import ChordPage, {loader} from '../ChordPage';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { describe, it, beforeEach, expect, vi } from 'vitest';
import SpellerApi from '../../../api';


vi.mock('../../../api',  () => {
    const mockChords = [
        {name: "C"},
        {name: "Cmaj7"},
        {name: "E"},
        {name: "F"},
    ];
    return {
        default: vi.fn().mockImplementation(() => {
            return {
                getAllChords: vi.fn().mockResolvedValue({chords: mockChords})                
            }
        })         
    }
});


describe('ChordPage component', () => {
    let router

  beforeEach(() => {
    localStorage.clear();
    
    const mockLoader = async () => {
        const apiInstance = new SpellerApi();
        const res = await apiInstance.getAllChords();
        console.log(res)
        return {chords: res.chords};
    }

    router = createMemoryRouter([
        {
            path: '/',
            element: <ChordPage/>,
            loader: mockLoader
        },
    ], {
        initialEntries: ['/'],
    }); 


  });

  it("matches snapshot", () => {
    const {asFragment} = render(<RouterProvider router={router} />)
    expect(asFragment()).toMatchSnapshot();
  })

  it("renders without crashing", () => {
    render(<RouterProvider router={router} />)
  })

  it("renders the autofill with the chords from database", async () => {
    console.log("running")

    localStorage.setItem('token', 'sampletoken')

    render(<RouterProvider router={router} />);

    await waitFor(() => {
        fireEvent.change(screen.getByLabelText("chord-input"), {target: {value: 'c'}})
        expect(screen.getByText("C")).toBeInTheDocument()
        expect(screen.getByText("Cmaj7")).toBeInTheDocument()
    })

  })
});