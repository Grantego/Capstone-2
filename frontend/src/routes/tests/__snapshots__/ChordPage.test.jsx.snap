// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChordPage component > matches snapshot 1`] = `<DocumentFragment />`;

exports[`Root component > matches snapshot 1`] = `
<DocumentFragment>
  <h2>
    Unexpected Application Error!
  </h2>
  <h3
    style="font-style: italic;"
  >
    Cannot read properties of undefined (reading 'map')
  </h3>
  <pre
    style="padding: 0.5rem; background-color: rgba(200, 200, 200, 0.5);"
  >
    TypeError: Cannot read properties of undefined (reading 'map')
    at ChordPage (/home/grantego/Springboard/Capstone-2/frontend/src/routes/ChordPage.jsx:19:41)
    at renderWithHooks (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:15486:18)
    at mountIndeterminateComponent (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:20103:13)
    at beginWork (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:21626:16)
    at beginWork$1 (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:27465:14)
    at performUnitOfWork (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:26599:12)
    at workLoopSync (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:26505:5)
    at renderRootSync (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:26473:7)
    at recoverFromConcurrentError (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:25889:20)
    at performConcurrentWorkOnRoot (/home/grantego/Springboard/Capstone-2/frontend/node_modules/react-dom/cjs/react-dom.development.js:25789:22)
  </pre>
  <p>
    ðŸ’¿ Hey developer ðŸ‘‹
  </p>
  <p>
    You can provide a way better UX than this when your app throws errors by providing your own 
    <code
      style="padding: 2px 4px; background-color: rgba(200, 200, 200, 0.5);"
    >
      ErrorBoundary
    </code>
     or 
    <code
      style="padding: 2px 4px; background-color: rgba(200, 200, 200, 0.5);"
    >
      errorElement
    </code>
     prop on your route.
  </p>
</DocumentFragment>
`;
